[v6r20] Run Pylint and create docs in Travis as well, pylint fixes

andresailer opened the pull request:
With this change unittests, pylint -E and docs are run in travis.

Note that `pylint -E` gives different errors than if just pylint is run. Don't ask me why...

BEGINRELEASENOTES

*Core
CHANGE: dropped scripts dirac-create-svn-tag, dirac-create-svn-branch

ENDRELEASENOTES

fstagni commented:
Now the pylint tests fail the PR, of course. It also gives quite some errors that should be ignored. I suggest to use:

pylint --rcfile=DIRAC/.pylintrc --errors-only --extension-pkg-whitelist=GSI,numpy

fstagni commented:
And 1 question: the docs creation is here only to check that we can actually build the docs?

andresailer commented:
I added the extension-pkg-whitelist. the `.pylintrc` is already in the command, with the correct path, because the working directory is DIRAC. There are less errors, but still some.

Yes, the docs are just build to ensure it is working. No more thread starting PRs breaking my docs :)

fstagni commented:
I fixed the errors in Graphs modules in #3497

andresailer commented:
Ok, thanks, I will not touch those here.

@atsareg , @iueda and others there are some pylint errors in the dirac-create-svn-* scripts are they still needed? (Is belleDirac still on svn? are those scripts used for making svn tags for extensions?)

andresailer commented:
I fixed a couple more pylint errors, some ignoring, some easy fixes.
The rest I don't know what to do with so please comment.

###  Module DIRAC.DataManagementSystem.Service.StorageElementProxyHandler
```
DataManagementSystem/Service/StorageElementProxyHandler.py:213: [E1126(invalid-sequence-index), StorageElementProxyHandler.export_prepareFileForHTTP] Sequence index is not an int, slice, or instance with __index__
```
No idea if this is a bug or pylint failing to figure out what is going on

###  Module dirac-admin-allow-se
```
DataManagementSystem/scripts/dirac-admin-allow-se.py:131: [E1101(no-member), ] Instance of 'ResourceStatus' has no 'setStorageElementStatus' member
```
fixed in #3502 

### Graphs

All fixed in #3497  ?

Module Test_Graph
```
Core/Utilities/test/Test_Graph.py:44: [E1101(no-member), DynamicPropTests.testDynamicProps] Instance of 'TestClass' has no 'makeProperty' member
Core/Utilities/test/Test_Graph.py:47: [E1101(no-member), DynamicPropTests.testDynamicProps] Instance of 'TestClass' has no 'rwTestProp' member
Core/Utilities/test/Test_Graph.py:50: [E1101(no-member), DynamicPropTests.testDynamicProps] Instance of 'TestClass' has no 'makeProperty' member
Core/Utilities/test/Test_Graph.py:93: [E1101(no-member), NodeTests.testNode] Instance of 'Node' has no 'visited' member
Core/Utilities/test/Test_Graph.py:116: [E1101(no-member), NodeTests.testNode] Instance of 'Edge' has no 'fromNode' member
Core/Utilities/test/Test_Graph.py:117: [E1101(no-member), NodeTests.testNode] Instance of 'Edge' has no 'toNode' member
Core/Utilities/test/Test_Graph.py:154: [E1101(no-member), EdgeTests.testEdge] Instance of 'Edge' has no 'visited' member
Core/Utilities/test/Test_Graph.py:173: [E1101(no-member), EdgeTests.testEdge] Instance of 'Edge' has no 'fromNode' member
Core/Utilities/test/Test_Graph.py:174: [E1101(no-member), EdgeTests.testEdge] Instance of 'Edge' has no 'toNode' member
```

Module DIRAC.Core.Utilities.Graphs.PlotBase
```
Core/Utilities/Graphs/PlotBase.py:8: [E0611(no-name-in-module), ] No name 'GraphData' in module 'DIRAC.Core.Utilities.Graphs'
Core/Utilities/Graphs/PlotBase.py:8: [E0401(import-error), ] Unable to import 'DIRAC.Core.Utilities.Graphs.GraphData' (unindent does not match any outer indentation level (<string>, line 155))
```
Module DIRAC.Core.Utilities.Graphs.Graph
```
Core/Utilities/Graphs/Graph.py:18: [E0611(no-name-in-module), ] No name 'GraphData' in module 'DIRAC.Core.Utilities.Graphs'
Core/Utilities/Graphs/Graph.py:18: [E0401(import-error), ] Unable to import 'DIRAC.Core.Utilities.Graphs.GraphData' (unindent does not match any outer indentation level (<string>, line 155))
Core/Utilities/Graphs/Graph.py:338: [E1101(no-member), Graph.writeGraph] Instance of 'FigureCanvasAgg' has no 'print_svg' member
```

Module DIRAC.Core.Utilities.Graphs.GraphData
```
Core/Utilities/Graphs/GraphData.py:155: [E0001(syntax-error), ] unindent does not match any outer indentation level
```

Module DIRAC.Core.Utilities.Graphs.PieGraph
```
Core/Utilities/Graphs/PieGraph.py:17: [E0611(no-name-in-module), ] No name 'GraphData' in module 'DIRAC.Core.Utilities.Graphs'
Core/Utilities/Graphs/PieGraph.py:17: [E0401(import-error), ] Unable to import 'DIRAC.Core.Utilities.Graphs.GraphData' (unindent does not match any outer indentation level (<string>, line 155))
```
Module DIRAC.Core.Utilities.Graphs.QualityMapGraph
```
Core/Utilities/Graphs/QualityMapGraph.py:15: [E0611(no-name-in-module), ] No name 'GraphData' in module 'DIRAC.Core.Utilities.Graphs'
Core/Utilities/Graphs/QualityMapGraph.py:15: [E0401(import-error), ] Unable to import 'DIRAC.Core.Utilities.Graphs.GraphData' (unindent does not match any outer indentation level (<string>, line 155))
```
Module DIRAC.Core.Utilities.Graphs.Legend
```
Core/Utilities/Graphs/Legend.py:17: [E0611(no-name-in-module), ] No name 'GraphData' in module 'DIRAC.Core.Utilities.Graphs'
Core/Utilities/Graphs/Legend.py:17: [E0401(import-error), ] Unable to import 'DIRAC.Core.Utilities.Graphs.GraphData' (unindent does not match any outer indentation level (<string>, line 155))
```
### Module DIRAC.Core.Workflow.Parameter
```
Core/Workflow/Parameter.py:410: [E1120(no-value-for-parameter), ParameterCollection.unlink] No value for argument 'opt' in method call
```
Bug
```
    elif isinstance( opt, Parameter ):
      self.unlink()
```
should be `opt.unlink()`?

###  Module DIRAC.Core.Workflow.Module
```
Core/Workflow/Module.py:150: [E1101(no-member), ModuleInstance.execute] Instance of 'ParameterCollection' has no 'execute' member
```
bug? No idea what this is supposed to do.

###  Module DIRAC.Core.Workflow.Step
```
Core/Workflow/Step.py:44: [E1121(too-many-function-args), StepDefinition.__init__] Too many positional arguments for constructor call
```
```
      self.parameters = ParameterCollection( self, obj.parameters )
```
Just `self.parameters = ParameterCollection( obj.parameters )`?

```
Core/Workflow/Step.py:110: [E1101(no-member), StepDefinition.updateParent] Class 'AttributeCollection' has no 'updateParents' member
```
bug? No idea what is going on...

###  Module DIRAC.WorkloadManagementSystem.Agent.PilotMonitorAgent
```
WorkloadManagementSystem/Agent/PilotMonitorAgent.py:15: [E0611(no-name-in-module), ] No name 'WMSAdministratorHandler' in module 'DIRAC.WorkloadManagementSystem.Client'
WorkloadManagementSystem/Agent/PilotMonitorAgent.py:15: [E0401(import-error), ] Unable to import 'DIRAC.WorkloadManagementSystem.Client.WMSAdministratorHandler'
```
Bug, WMSAdministratorHandler is in Service, but one has to make an RPC client here

###  Module DIRAC.Workflow.Modules.ModuleBase
```
Workflow/Modules/ModuleBase.py:344: [E1101(no-member), ModuleBase._determineStepInputData] Instance of 'ModuleBase' has no 'inputDataType' member
```
bug, but again no idea what this should be

fstagni commented:
First: thanks a lot, great job!

- StorageElementProxyHandler: I don't know, really. I saw before and I had your same doubts. Maybe @chaen or @atsareg can answer 
- Graphs modules should be all fixed in the other PR of mine, I think. They are bugs introduced within the last tag, so it should be OK.
- For the DIRAC.Core.Workflow stuff: I never dared to try and fix them, as they are definitely central modules. These modules have also zillions of other pylint warnings which should be addressed at some point. They are, also, over complicated and I have the impression that a decent part of the code in there is never used anywhere. On the specific errors:

1. Parameter.py: I also guess it should be `opt.unlink()`
2. Module.py: no idea. I'd put a `pylint:disable` together with `FIXME` note.
3. Step.py:44 : I guess it's as you are saying. Anyway it probably means that the `elif` at line 37 is dead code.
4. Step.py:110: no idea either. Maybe also here I'd put a `pylint:disable` together with `FIXME` note.

- PilotMonitorAgent: I don't see these errors... why?
- ModuleBase: add the following lines in the module: `self.inputDataType = None` in the `__init__` and `self.inputDataType = self.step_commons.get( 'inputDataType', self.inputDataType )` in the `_resolveInputStep`



Anyway: I suggest to wait before merging this PR until at least #3497 is merged, then you rebase and only then we merge, if everything's OK.

fstagni commented on WorkloadManagementSystem/DB/PilotsLoggingDB.py:
```diff
@@ -77,7 +77,7 @@ def __initializeDB( self ):
     if not 'PilotsLogging' in tablesInDB:
       try:
-        PilotsLogging.__table__.create( self.engine )
+        PilotsLogging.__table__.create( self.engine ) #pylint: disable=no-member
```
Please remove this one: I fixed, together with some more fixes, in #3497

andresailer replied:
Dropped commit

andresailer commented:
All done as you say. 
Except `self.inputDataType = ''` i.e. a string, otherwise the `.lower()` gives exception for `None`.

iueda commented:
> @atsareg , @iueda and others there are some pylint errors in the dirac-create-svn-* scripts are they still needed? (Is belleDirac still on svn? are those scripts used for making svn tags for extensions?)

We are not using SVN for BelleDIRAC any longer, and not using those scripts.

fstagni commented:
OK thanks. So I think we can mark all the other "SVN" stuff as deprecated.

fstagni commented:
Do you want to try to merge or rebase on integration right now?

andresailer commented:
I will rebase

andresailer commented:
Still some pylint errors in the Graph, its tests, and the StorageElementProxyHandler

I will add ignores for Test_Graph and check about the `print_svg`.
still waiting for @chaen and @atsareg about the StorageElementProxyHandler. Is this function ever called? Then it would be a false positive.

```
************* Module DIRAC.DataManagementSystem.Service.StorageElementProxyHandler

DataManagementSystem/Service/StorageElementProxyHandler.py:213: [E1126(invalid-sequence-index), StorageElementProxyHandler.export_prepareFileForHTTP] Sequence index is not an int, slice, or instance with __index__

************* Module Test_Graph

Core/Utilities/test/Test_Graph.py:44: [E1101(no-member), DynamicPropTests.testDynamicProps] Instance of 'TestClass' has no 'makeProperty' member

Core/Utilities/test/Test_Graph.py:47: [E1101(no-member), DynamicPropTests.testDynamicProps] Instance of 'TestClass' has no 'rwTestProp' member

Core/Utilities/test/Test_Graph.py:50: [E1101(no-member), DynamicPropTests.testDynamicProps] Instance of 'TestClass' has no 'makeProperty' member

Core/Utilities/test/Test_Graph.py:93: [E1101(no-member), NodeTests.testNode] Instance of 'Node' has no 'visited' member

Core/Utilities/test/Test_Graph.py:116: [E1101(no-member), NodeTests.testNode] Instance of 'Edge' has no 'fromNode' member

Core/Utilities/test/Test_Graph.py:117: [E1101(no-member), NodeTests.testNode] Instance of 'Edge' has no 'toNode' member

Core/Utilities/test/Test_Graph.py:154: [E1101(no-member), EdgeTests.testEdge] Instance of 'Edge' has no 'visited' member

Core/Utilities/test/Test_Graph.py:173: [E1101(no-member), EdgeTests.testEdge] Instance of 'Edge' has no 'fromNode' member

Core/Utilities/test/Test_Graph.py:174: [E1101(no-member), EdgeTests.testEdge] Instance of 'Edge' has no 'toNode' member

************* Module DIRAC.Core.Utilities.Graphs.Graph

Core/Utilities/Graphs/Graph.py:338: [E1101(no-member), Graph.writeGraph] Instance of 'FigureCanvasAgg' has no 'print_svg' member
```

chaen commented:
Sorry, did not see. I will check now

fstagni commented:
Maybe @zmathe knows about the Graph modules errors? They come from matplotlib.

andresailer commented:
I think the `print_svg` is a real bug, because one would need a different backend (https://matplotlib.org/api/backend_svg_api.html#matplotlib.backends.backend_svg.FigureCanvasSVG ), but maybe there is some magic pylint and I are missing

fstagni commented:
Or maybe the object used is not of type `FigureCanvasSVG`

chaen commented:
OK, got it !! (if you've never used pdb against pylint, you should :-) ). 
TLDR: we can ignore.

Long explanation if people are interested.

Basically, this is the part we are interested in:

```
    result = pythonCall( 300, self.__prepareFileForHTTP, lfn, key )
    if result['OK']:
      result = result['Value']
      if result['OK']:
```

Pylint says ```Sequence index is not an int, slice, or instance with __index__``` about the last ```if``` statement

With the debugger, I understood that pylint was convinced that at this level, `result` was a tuple. 
Indeed, we can reproduce easily:

```
[chaen@pclhcb31 tmp]$ cat subList.py 
res = (1,2,3)
print res['Value']

[chaen@pclhcb31 tmp]$ pylint subList.py | grep Sequence
E:  2, 0: Sequence index is not an int, slice, or instance with __index__ (invalid-sequence-index)
```

The details of the reason why it believes it to be a tuple is unknown to me (I do not want to check their lexical tree), but it comes from there:

```
ipdb> p parent_type
<Tuple(tuple) l.136 [DIRAC.Core.Utilities.Subprocess] at 0x7f961ca7e790>
```
In Subprocess.py
```
    except Exception, error:
      return { "OK" : False, "Message" : str(error),
               "Value": ( 2, '', '' )  }
```

This is the only type that pylint can infer, since all the rest is dynamic and duck typing, so I suppose it falls back on that. 

Andre, can you please add a disable statement on this specific line, with a comment along the line `believes is a tuple because it is the only possible return type pylint finds in Subprocess.py` ?

andresailer commented:
Added the ignore and comment from @chaen to StorageElementHandlerProxy.
Add ignores to Test_Graph, more metaclass things pylint does not understand.

zmathe commented:
print_svg is a real bug. We do not see it, because we always save PNG. 
```python
def writeGraph( self, fname, fileFormat = 'PNG' ):
    """ Write out the resulting graph to a file with fname in a given format
    """

    self.canvas.draw()
    if fileFormat.lower() == 'png':
      self.canvas.print_png(fname)
    elif fileFormat.lower() == 'svg':
      self.canvas.print_svg(fname)

```
I suggest to remove:
```python
 elif fileFormat.lower() == 'svg':
      self.canvas.print_svg(fname)
```
Anyway, we will be not able to save svg using this backend!

andresailer commented:
Tests should be consistently passing now. Please review

fstagni commented:
Why extending the timeout?

andresailer commented:
Because the find+pylint doesn;t produce any output so travis considers the task stalled if it takes longer than 10 minutes, which happened on my branch https://travis-ci.org/andresailer/DIRAC/jobs/294747143#L707

andresailer commented:
coveralls was not working because
```
11.79s$ coveralls
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.13/bin/coveralls", line 11, in <module>
    sys.exit(wear())
  File "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/coveralls/__init__.py", line 95, in wear
    source_files=coverage.coveralls(args.base_dir, ignore_errors=args.ignore_errors, merge_file=args.merge_file),
  File "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/coveralls/control.py", line 9, in coveralls
    return reporter.report(base_dir, ignore_errors=ignore_errors, merge_file=merge_file)
  File "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/coveralls/report.py", line 15, in report
    with open(fr.filename) as fp:
IOError: [Errno 2] No such file or directory: u'/home/travis/build/andresailer/DIRAC/testPoolCEJob_0.py'
```
Added flag to ignore errors when reading source files

zmathe commented on Core/Utilities/Graphs/Graph.py:
```diff
@@ -334,5 +334,3 @@ def writeGraph( self, fname, fileFormat = 'PNG' ):
     self.canvas.draw()
     if fileFormat.lower() == 'png':
       self.canvas.print_png(fname)
-    elif fileFormat.lower() == 'svg':
```
Could you please add an error message?
```python
if fileFormat.lower() == 'png':
  self.canvas.print_png(fname)
else:
  gLogger.error("%s is not supported file format" %fileFormat)
```
In case someone will modify it will get an error message. Thanks.

andresailer replied:
Done

fstagni commented:
I would be in favor of merging this PR directly, then rebase other existing PRs. Any objection?

andresailer commented:
@atsareg or any one else have any objection to have this merged?

fstagni commented:
Merging in 3...2...1 (hour)

fstagni commented:
I will now update the other PRs (merge integration branch).
